// webpack-config/dist/index.js
'use strict';

Object.defineProperty(exports, "__esModule", {
  value: true
});
exports.optionsResolver = exports.factory = exports.finder = exports.loader = exports.pathResolver = exports.stringResolver = exports.patternCache = exports.cache = exports.environment = exports.ConfigOptionsResolver = exports.ConfigFactory = exports.ConfigPathResolver = exports.ConfigStringResolver = exports.ConfigEnvironment = exports.ConfigFinder = exports.ConfigLoader = exports.ConfigCache = exports.ConfigPatternCache = exports.ConfigBuilder = exports.Config = undefined;

var _Config = require('./Config');

var _Config2 = _interopRequireDefault(_Config);

var _ConfigBuilder = require('./ConfigBuilder');

var _ConfigBuilder2 = _interopRequireDefault(_ConfigBuilder);

var _ConfigCache = require('./ConfigCache');

var _ConfigCache2 = _interopRequireDefault(_ConfigCache);

var _ConfigLoader = require('./ConfigLoader');

var _ConfigLoader2 = _interopRequireDefault(_ConfigLoader);

var _ConfigFinder = require('./ConfigFinder');

var _ConfigFinder2 = _interopRequireDefault(_ConfigFinder);

var _ConfigEnvironment = require('./ConfigEnvironment');

var _ConfigEnvironment2 = _interopRequireDefault(_ConfigEnvironment);

var _ConfigPatternCache = require('./ConfigPatternCache');

var _ConfigPatternCache2 = _interopRequireDefault(_ConfigPatternCache);

var _ConfigStringResolver = require('./ConfigStringResolver');

var _ConfigStringResolver2 = _interopRequireDefault(_ConfigStringResolver);

var _ConfigPathResolver = require('./ConfigPathResolver');

var _ConfigPathResolver2 = _interopRequireDefault(_ConfigPathResolver);

var _ConfigFactory = require('./ConfigFactory');

var _ConfigFactory2 = _interopRequireDefault(_ConfigFactory);

var _ConfigContainer = require('./ConfigContainer');

var _ConfigContainer2 = _interopRequireDefault(_ConfigContainer);

var _ConfigOptionsResolver = require('./ConfigOptionsResolver');

var _ConfigOptionsResolver2 = _interopRequireDefault(_ConfigOptionsResolver);

function _interopRequireDefault(obj) { return obj && obj.__esModule ? obj : { default: obj }; }

/**
 * @private
 * @type {ConfigContainer}
 */
var container = new _ConfigContainer2.default();

/**
 * Proxy class which automatically fills {@link Config} constructor dependencies
 * @class
 * @extends {Config}
 */
var ConfigProxy = container.proxy(_Config2.default);

/**
 * Proxy class which automatically fills {@link ConfigBuilder} constructor dependencies
 * @class
 * @extends {ConfigBuilder}
 */
var ConfigBuilderProxy = container.proxy(_ConfigBuilder2.default);

/**
 * @module webpack-config
 */

exports.default = ConfigProxy;
exports.Config = ConfigProxy;
exports.ConfigBuilder = ConfigBuilderProxy;
exports.

/**
 * @type {ConfigPatternCache}
 */
ConfigPatternCache = _ConfigPatternCache2.default;
exports.

/**
 * @type {ConfigCache}
 */
ConfigCache = _ConfigCache2.default;
exports.

/**
 * @type {ConfigLoader}
 */
ConfigLoader = _ConfigLoader2.default;
exports.

/**
 * @type {ConfigFinder}
 */
ConfigFinder = _ConfigFinder2.default;
exports.

/**
 * @type {ConfigEnvironment}
 */
ConfigEnvironment = _ConfigEnvironment2.default;
exports.

/**
 * @type {ConfigStringResolver}
 */
ConfigStringResolver = _ConfigStringResolver2.default;
exports.

/**
 * @type {ConfigPathResolver}
 */
ConfigPathResolver = _ConfigPathResolver2.default;
exports.

/**
 * @type {ConfigFactory}
 */
ConfigFactory = _ConfigFactory2.default;
exports.

/**
 * @type {ConfigOptionsResolver}
 */
ConfigOptionsResolver = _ConfigOptionsResolver2.default;

/**
 * @type {ConfigEnvironment}
 */

var environment = exports.environment = container.resolve(_ConfigEnvironment2.default);

/**
 * @type {ConfigCache}
 */
var cache = exports.cache = container.resolve(_ConfigCache2.default);

/**
 * @type {ConfigPatternCache}
 */
var patternCache = exports.patternCache = container.resolve(_ConfigPatternCache2.default);

/**
 * @type {ConfigStringResolver}
 */
var stringResolver = exports.stringResolver = container.resolve(_ConfigStringResolver2.default);

/**
 * @type {ConfigPathResolver}
 */
var pathResolver = exports.pathResolver = container.resolve(_ConfigPathResolver2.default);

/**
 * @type {ConfigLoader}
 */
var loader = exports.loader = container.resolve(_ConfigLoader2.default);

/**
 * @type {ConfigFinder}
 */
var finder = exports.finder = container.resolve(_ConfigFinder2.default);

/**
 * @type {ConfigFactory}
 */
var factory = exports.factory = container.resolve(_ConfigFactory2.default);

/**
 * @type {ConfigOptionsResolver}
 */
var optionsResolver = exports.optionsResolver = container.resolve(_ConfigOptionsResolver2.default);
